### ISSUE TITLE
Test prepared statements for MySQL

### ISSUE THREAD

**Issue by jamadden** [zodb/relstorage#156]
Similar to #148.

There are limits under MySQL, though. Most notably, unless the driver supports the binary protocol for prepared statements (only the official MySQL Connector/Python does), each parameter to the statement must be sent as a user variable ('SET x = ...') which implies an additional trip to the server. (There doesn't seem to be any sort of buffering or grouping unless it happens at the Nagle TCP_NDELAY level.) So the benefits need to be tested to see if its worth it.

They could still be useful for the statements that take zero parameters, like the poll statement.

Another *possible* problem is that [there is a limit of 16382](http://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html#sysvar_max_prepared_stmt_count) outstanding prepared statements per server by default. Suppose you have three sites sharing a server, each with three databases, 20 workers per instance/site and 2 instances/site. Thats `(16382)/(3 * 3 * 20 * 2) = 45` prepared statements allowed at most.

**Comment by jamadden** [comment]
Unfortunately, for the main `load_current` query, the overhead of executing the additional server round trip absolutely destroys performance. To take one example, the "cold" zodbshootout number started at 15,665 (my machine, unix-socket connection, mysql 5.5). Going to a prepared statement with one variable dropped that number to 11,368, or almost *28%*! Using TCP only makes it worse.

Ahh, but there's a light at the end of the tunnel. Stored procedures can be accessed with the `CALL` statement, and the `CALL` statement will accept literals (`CALL load_current(0)`). So we eliminate the network round trip, right? No, the light turned out to be an oncoming train: calling a stored procedure always generates *two* result sets (the one from the `SELECT` statement in the procedure and the one from the `CALL` itself). That means you have to use `cursor.nextset()` to avoid getting the protocol out of sync. And, you guessed it, `cursor.nextset()` requires a trip to the server. The `CALL` does perform a bit better, though: 13,101 for a drop of only 17%.

**Comment by jamadden** [comment]
Of course, that was with the default size of 96 bytes per object. I need to check on other sizes to see if/when there's a breakeven point. For comparison, one production database I looked at that has 18.6 million objects in it had an average object size of 244 (stddev 392), max of 73700 and min of 21 (those sizes are compressed by zc.zlibstorage).

**Comment by jamadden** [comment]
Even using objects as large as 1024, the performance difference is still there. It shrinks some (down to about 13%) but its still there. Conclusion: prepared statements/stored procedures are not worth using for load_current or other parameterized queries in MySQL.

There are still a number of possible other queries that might benefit though.

**Comment by jamadden** [comment]
I've finished evaluating the common queries.

I also tested the binary protocol using MySQL Connector/Python which (theoretically) can eliminate a round trip. At least in that implementation, [it wasn't worth it](https://github.com/zodb/relstorage/issues/155#issuecomment-275414459). 

I'm going to close this now, we can reopen if anything changes (for example, if mysqlclient or PyMySQL supports the binary protocol).
